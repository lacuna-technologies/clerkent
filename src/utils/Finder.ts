import Constants from './Constants'
import type Law from '../types/Law'
export interface FinderResult {
  jurisdiction: Law.JursidictionCode
  citation: string,
  index: number,
  court? : string
}

// const emptyParseResult: ParseResult = {
//   citation: undefined,
//   jurisdiction: undefined,
// }

// const parseQuery = (query: string): ParseResult => {
//   const cleanedQuery = query.trim()
//   return parseSGCase(cleanedQuery) || emptyParseResult
// }

// const inSLW = (query: string) => {
//   const eligibleCourt = Array.isArray(query.match(/SGCA|SGHC/))

//   const yearRegex = /^\[(2\d{3})]/
//   const yearMatch = query.match(yearRegex)
//   const eligibleYear = Array.isArray(yearMatch) && Number.parseInt(yearMatch[1]) >= 2000

//   return eligibleCourt && eligibleYear
// }

// const inSGSC = (query: string) => inSLW(query)

const findCase = (query: string): FinderResult[] => {
  return [
    ...findSGCase(query),
    ...findUKCase(query),
    ...findEUCase(query),
    ...findHKCase(query),
  ]
}

const findSGCase = (query: string): FinderResult[] => {
  const regex = /\[[12]\d{3}]( \d{1,2})? (sgca|sghc|sgdc|sgmc|slr(\(r\))?) \d{1,4}/gi
  const matches = [...query.matchAll(regex)]
  if (matches.length > 0) {
    return matches.map((match) => ({
      citation: match[0],
      index: match.index,
      jurisdiction: Constants.JURISDICTIONS.SG.id,
    }))
  }
  return []
}

const findUKCase = (query:string): FinderResult[] => {
  const abbrs = [
    {
      abbr: `EWCA`,
      appendum: `( Civ)?`,
    },
    {
      abbr: `EWHC`,
      appendum: `( Patents)?`,
    },
    {
      abbr: `UKSC`,
    },
    {
      abbr: `UKPC`,
    },
    {
      abbr: `UKHL`,
    },
    {
      abbr: `AC`,
    },
    {
      abbr: `UKHL`,
    },
    {
      abbr: `Ch`,
      appendum: `( D)?`,
    },
    {
      abbr: `QB`,
      appendum: `(D)?`,
    },
    {
      abbr: `KB`,
    },
    {
      abbr: `WLR`,
      appendum: `( \\(D\\))?`,
    },
    {
      abbr: `All ER`,
      appendum: `( \\(D\\))?`,
    },
    {
      abbr: `BCLC`,
    },
    {
      abbr: `BCC`,
    },
    {
      abbr: `HL Cas`,
    },
    {
      abbr: `App Cas`,
    },
    {
      abbr: `Ld Raym`,
    },
    {
      abbr: `FSR`,
    },
    {
      abbr: `ECC`,
    },
    {
      abbr: `ITCLR`,
    },
    {
      abbr: `RPC`,
    },
    {
      abbr: `Ex Rep`,
    },
    {
      abbr: `ER`,
    },
  ].map(({ abbr, appendum }) => `${abbr
    .split(``)
    .map(letter =>
      /[a-z]/i.test(letter)
        ? letter+`\\.?`
        : letter,
    ).join(``)
    }${appendum ? appendum : ``}`,
  ).join(`|`)
  // eslint-disable-next-line unicorn/better-regex
  const yearRegex = new RegExp(/((\[|\()[12]\d{3}(-[12]\d{3})?(\]|\)))/)
  const volumeRegex = new RegExp(/( \d{1,2})?/)
  const pageRegex = new RegExp(/\d{1,4}/)
  const regex = new RegExp(`${yearRegex.source}${volumeRegex.source} (${abbrs}) ${pageRegex.source}`, `gi`)

  const matches = [...query.matchAll(regex)]
  if (matches.length > 0) {
    return matches.map((match) => ({
      citation: match[0],
      index: match.index,
      jurisdiction: Constants.JURISDICTIONS.UK.id,
    }))
  }
  return []
}

export const epoRegex = new RegExp(/T ?\d{1,4}\/\d{1,2}/)
export const cjeuRegex = new RegExp(/[CT]-\d{1,3}\/\d{1,2}/)

const findEUCase = (query: string): FinderResult[] => {
  const regex = new RegExp(`(${epoRegex.source})|(${cjeuRegex.source})`, `gi`)
  const cleanedQuery = query.replaceAll(`â€‘`, `-`)

  const matches = [...cleanedQuery.matchAll(regex)]
  if(matches.length > 0){
    return matches.map((match) => ({
      citation: match[0],
      court: match[1] ? Constants.COURTS.EU_epo.id : Constants.COURTS.EU_cjeu.id,
      index: match.index,
      jurisdiction: Constants.JURISDICTIONS.EU.id,
    }))
  }
  return []
}

const findHKCase = (query: string): FinderResult[] => {
  const regex = /\[[12]\d{3}] (HKCA|HKCFA|HKCFI) \d{1,4}/g
  const matches = [...query.matchAll(regex)]
  if (matches.length > 0) {
    return matches.map((match) => ({
      citation: match[0],
      index: match.index,
      jurisdiction: Constants.JURISDICTIONS.HK.id,
    }))
  }
  return []
}

const Finder = {
  findCase,
  findEUCase,
  findHKCase,
  findSGCase,
  findUKCase,
}

export default Finder