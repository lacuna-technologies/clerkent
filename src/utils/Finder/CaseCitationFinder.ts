import Constants from '../Constants'
import type Law from '../../types/Law'

export interface CaseCitationFinderResult {
  jurisdiction: Law.JursidictionCode
  citation: string,
  index: number,
  court? : string,
  type: `case-citation`
}

// const emptyParseResult: ParseResult = {
//   citation: undefined,
//   jurisdiction: undefined,
// }

// const parseQuery = (query: string): ParseResult => {
//   const cleanedQuery = query.trim()
//   return parseSGCase(cleanedQuery) || emptyParseResult
// }

// const inSLW = (query: string) => {
//   const eligibleCourt = Array.isArray(query.match(/SGCA|SGHC/))

//   const yearRegex = /^\[(2\d{3})]/
//   const yearMatch = query.match(yearRegex)
//   const eligibleYear = Array.isArray(yearMatch) && Number.parseInt(yearMatch[1]) >= 2000

//   return eligibleCourt && eligibleYear
// }

// const inSGSC = (query: string) => inSLW(query)

const formatAbbrs = (abbrArray) => abbrArray.map(({ abbr, appendum }) => `${abbr
    .split(``)
    .map((letter: string) =>
      /[a-z]/i.test(letter)
        ? letter+`\\.?`
        : letter,
    ).join(``)
    }${appendum ? appendum : ``}`,
  ).join(`|`)

const findSGCaseCitation = (query: string): CaseCitationFinderResult[] => {
  const regex = /\[[12]\d{3}]( \d{1,2})? (sgca(\(i\))?|sghc|sgdc|sgmc|slr(\(r\))?) \d{1,4}/gi
  const matches = [...query.matchAll(regex)]
  if (matches.length > 0) {
    return matches.map((match) => ({
      citation: match[0],
      index: match.index,
      jurisdiction: Constants.JURISDICTIONS.SG.id,
    })).map(c => ({ ...c, type: `case-citation` }))
  }
  return []
}

const findUKCaseCitation = (query:string): CaseCitationFinderResult[] => {
  const abbrs = formatAbbrs([
    { abbr: `EWCA`, appendum: `( Civ| Crim)?` },
    { abbr: `EWHC`, appendum: `( Patents)?` },
    { abbr: `UKSC` },
    { abbr: `UKPC` },
    { abbr: `UKHL` },
    { abbr: `AC` },
    { abbr: `Ch`, appendum: `( D)?` },
    { abbr: `QB`, appendum: `(D)?` },
    { abbr: `KB` },
    { abbr: `WLR`, appendum: `( ?\\(D\\))?` },
    { abbr: `All ER`, appendum: `( \\((D|Comm.?)\\))?` },
    { abbr: `BCLC` },
    { abbr: `BCC` },
    { abbr: `HL Cas` },
    { abbr: `App Cas` },
    { abbr: `Ld Raym` },
    { abbr: `FSR` },
    { abbr: `ECC` },
    { abbr: `ITCLR` },
    { abbr: `RPC` },
    { abbr: `Ex Rep` },
    { abbr: `ER` },
    { abbr: `Cr App R` },
    { abbr: `ALR` },
    { abbr: `FLR` },
    { abbr: `Hare` },
    { abbr: `H & Tw` },
    { abbr: `EMLR` },
    { abbr: `Fam` },
    { abbr: `Macq` },
    { abbr: `TLR` },
    { abbr: `Ves & B` },
    { abbr: `EngR` },
    { abbr: `Lloyd's Rep` },
    { abbr: `BLR` },
    { abbr: `CLC` },
  ])
  // eslint-disable-next-line unicorn/better-regex
  const yearRegex = new RegExp(/((\[|\()[12]\d{3}(-[12]\d{3})?(\]|\)))/)
  const volumeRegex = new RegExp(/( \d{1,2})?/)
  const pageRegex = new RegExp(/\d{1,4}/)
  const regex = new RegExp(`${yearRegex.source}${volumeRegex.source} (${abbrs}) ${pageRegex.source}`, `gi`)

  const matches = [...query.matchAll(regex)]
  if (matches.length > 0) {
    return matches.map((match) => ({
      citation: match[0],
      index: match.index,
      jurisdiction: Constants.JURISDICTIONS.UK.id,
    })).map(c => ({ ...c, type: `case-citation` }))
  }
  return []
}

export const epoRegex = new RegExp(/[GJT][ _]?\d{1,4}\/\d{1,2}/)
export const cjeuRegex = new RegExp(/[CT]-\d{1,3}\/\d{1,2}/)

const findEUCaseCitation = (query: string): CaseCitationFinderResult[] => {
  const regex = new RegExp(`(${epoRegex.source})|(${cjeuRegex.source})`, `gi`)
  const cleanedQuery = query.replace(/â€‘/g, `-`)

  const matches = [...cleanedQuery.matchAll(regex)]
  if(matches.length > 0){
    return matches.map((match) => ({
      citation: match[0],
      court: match[1] ? Constants.COURTS.EU_epo.id : Constants.COURTS.EU_cjeu.id,
      index: match.index,
      jurisdiction: Constants.JURISDICTIONS.EU.id,
    })).map(c => ({ ...c, type: `case-citation` }))
  }
  return []
}

const findHKCaseCitation = (query: string): CaseCitationFinderResult[] => {
  // eslint-disable-next-line unicorn/better-regex
  const yearRegex = new RegExp(/(([([])[12]\d{3}(-[12]\d{3})?[)\]])/)
  const volumeRegex = new RegExp(/( \d{1,2})?/)
  const pageRegex = new RegExp(/\d{1,4}/)
  const abbrs = formatAbbrs([
    { abbr: `HKCA` },
    { abbr: `HKCFA` },
    { abbr: `HKCFI` },
    { abbr: `HKLRD` },
    { abbr: `HKCFAR` },
    { abbr: `HKC` },
    { abbr: `HKDC` },
    { abbr: `HKFC` },
    { abbr: `HKLdT` },
    { abbr: `HKCT` },
    { abbr: `HKCrC` },
    { abbr: `HKOAT` },
    { abbr: `HKLaT` },
    { abbr: `HKMagC` },
  ])
  const regex = new RegExp(`${yearRegex.source}${volumeRegex.source} (${abbrs}) ${pageRegex.source}`, `gi`)

  const matches = [...query.matchAll(regex)]
  if (matches.length > 0) {
    return matches.map((match) => ({
      citation: match[0],
      index: match.index,
      jurisdiction: Constants.JURISDICTIONS.HK.id,
    })).map(c => ({ ...c, type: `case-citation` }))
  }
  return []
}

const findCACaseCitation = (query: string): CaseCitationFinderResult[] => {
  // eslint-disable-next-line unicorn/better-regex
  const yearRegex = new RegExp(/(([([])[12]\d{3}(-[12]\d{3})?[)\]])/)
  const volumeRegex = new RegExp(/( \d{1,2})?/)
  const pageRegex = new RegExp(/\d{1,4}/)
  const abbrs = formatAbbrs([
    { abbr: `SCR` },
    { abbr: `WWR` },
    { abbr: `EXP` },
    { abbr: `CarswellAlta` },
    { abbr: `SCJ No` },
  ])
  const regex = new RegExp(`${yearRegex.source}${volumeRegex.source} (${abbrs}) ${pageRegex.source}`, `gi`)

  const matches = [...query.matchAll(regex)]
  if (matches.length > 0) {
    return matches.map((match) => ({
      citation: match[0],
      index: match.index,
      jurisdiction: Constants.JURISDICTIONS.CA.id,
    })).map(c => ({ ...c, type: `case-citation` }))
  }
  return []
}

const findAUCaseCitation = (query: string): CaseCitationFinderResult[] => {
  // eslint-disable-next-line unicorn/better-regex
  const yearRegex = new RegExp(/(([([])[12]\d{3}(-[12]\d{3})?[)\]])/)
  const volumeRegex = new RegExp(/( \d{1,2})?/)
  const pageRegex = new RegExp(/\d{1,4}/)
  const abbrs = formatAbbrs([
    { abbr: `HCA` },
    { abbr: `UKPCHCA` },
    { abbr: `FamCA` },
    { abbr: `HCASum` },
    { abbr: `NSWSC` },
    { abbr: `NSWCA` },
    { abbr: `NSWCCA` },
    { abbr: `NSWCIMC` },
    { abbr: `NSWCC` },
    { abbr: `NSWDC` },
    { abbr: `NSWDRGC` },
    { abbr: `NSWIC` },
    { abbr: `NSWKnoxRp` },
    { abbr: `NSWLR` },
    { abbr: `NSWR` },
    { abbr: `SR \\(NSW\\)`},
    { abbr: `CLR` },
    { abbr: `FCA` },
    { abbr: `FCAFC` },
    { abbr: `FCCA` },
    { abbr: `FMCA` },
    { abbr: `FMCAfam` },
    { abbr: `IRCA` },
    { abbr: `FCR` },
    { abbr: `ALR` },
    { abbr: `QCA` },
    { abbr: `QSC` },
    { abbr: `QDC` },
    { abbr: `QChC` },
    { abbr: `QChCM` },
    { abbr: `ICQ` },
    { abbr: `QIC` },
    { abbr: `QLC` },
    { abbr: `VSC` },
    { abbr: `VicSC` },
    { abbr: `VSCA` },
    { abbr: `VicCorC` },
    { abbr: `VCC` },
    { abbr: `VMC` },
    { abbr: `VicRp` },
    { abbr: `VicLawRp` },
    { abbr: `WASC` },
    { abbr: `WASCA` },
    { abbr: `WALawRp` },
    { abbr: `WADC` },
    { abbr: `FWCA` },
    { abbr: `FCWAM` },
    { abbr: `WACIC` },
    { abbr: `WAGAB` },
    { abbr: `WASupC` },
    { abbr: `WAR` },
    { abbr: `TASSC` },
    { abbr: `TASFC` },
    { abbr: `TASCCA` },
    { abbr: `TASMC` },
    { abbr: `TASSupC` },
    { abbr: `TASLawRp` },
    { abbr: `TASStRp` },
    { abbr: `TASRp` },
    { abbr: `TASADT` },
    { abbr: `TASFPT` },
    { abbr: `SASC` },
    { abbr: `SASCA` },
    { abbr: `SASCFC` },
    { abbr: `SADC` },
    { abbr: `SAERDC` },
    { abbr: `SALC` },
    { abbr: `SAPelhamRp` },
    { abbr: `SAIRC` },
    { abbr: `SAIndRp` },
    { abbr: `SALawRp` },
    { abbr: `SAStRp` },
    { abbr: `SAWC` },
    { abbr: `SACAT` },
    { abbr: `NTSC` },
    { abbr: `NTCA` },
    { abbr: `NTCCA` },
    { abbr: `NTMC` },
    { abbr: `NTJud` },
    { abbr: `NTLC` },
    { abbr: `ACTCA` },
    { abbr: `ACTSCFC` },
    { abbr: `ACTSC` },
    { abbr: `ACTCD` },
    { abbr: `ACTIC` },
    { abbr: `ACTMC` },
    { abbr: `ACAT` },
    { abbr: `ADO` },
    { abbr: `AUDND` },
    { abbr: `NFSC` },
  ])
  const regex = new RegExp(`${yearRegex.source}${volumeRegex.source} (${abbrs}) ${pageRegex.source}`, `gi`)

  const matches = [...query.matchAll(regex)]
  if (matches.length > 0) {
    return matches.map((match) => ({
      citation: match[0],
      index: match.index,
      jurisdiction: Constants.JURISDICTIONS.AU.id,
    })).map(c => ({ ...c, type: `case-citation` }))
  }
  return []
}

const findNZCaseCitation = (query: string): CaseCitationFinderResult[] => {
  // eslint-disable-next-line unicorn/better-regex
  const yearRegex = new RegExp(/(([([])[12]\d{3}(-[12]\d{3})?[)\]])/)
  const volumeRegex = new RegExp(/( \d{1,2})?/)
  const pageRegex = new RegExp(/\d{1,4}/)
  const abbrs = formatAbbrs([
    { abbr: `NZCA` },
    { abbr: `NZSC` },
    { abbr: `NZHC` },
    { abbr: `NZLR` },
    { abbr: `NZAR` },
    { abbr: `NZFC` },
    { abbr: `NZPC` },
    { abbr: `NZDC` },
  ])
  const regex = new RegExp(`${yearRegex.source}${volumeRegex.source} (${abbrs}) ${pageRegex.source}`, `gi`)

  const matches = [...query.matchAll(regex)]
  if (matches.length > 0) {
    return matches.map((match) => ({
      citation: match[0],
      index: match.index,
      jurisdiction: Constants.JURISDICTIONS.NZ.id,
    })).map(c => ({ ...c, type: `case-citation` }))
  }
  return []
}
const findCaseCitation = (query: string): CaseCitationFinderResult[] => {
  return [
    ...findSGCaseCitation(query),
    ...findUKCaseCitation(query),
    ...findEUCaseCitation(query),
    ...findHKCaseCitation(query),
    ...findCACaseCitation(query),
    ...findAUCaseCitation(query),
    ...findNZCaseCitation(query),
  ]
}

const CaseFinder = {
  findAUCaseCitation,
  findCACaseCitation,
  findCaseCitation,
  findEUCaseCitation,
  findHKCaseCitation,
  findSGCaseCitation,
  findUKCaseCitation,
}

export default CaseFinder