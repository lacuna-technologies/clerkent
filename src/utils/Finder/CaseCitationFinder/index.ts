export { default } from './CaseCitationFinder'
export type { CaseCitationFinderResult } from './types'