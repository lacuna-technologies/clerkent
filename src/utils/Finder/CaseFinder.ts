import Constants from '../Constants'
import type Law from '../../types/Law'
export interface CaseFinderResult {
  jurisdiction: Law.JursidictionCode
  citation: string,
  index: number,
  court? : string,
  type: `case`
}

// const emptyParseResult: ParseResult = {
//   citation: undefined,
//   jurisdiction: undefined,
// }

// const parseQuery = (query: string): ParseResult => {
//   const cleanedQuery = query.trim()
//   return parseSGCase(cleanedQuery) || emptyParseResult
// }

// const inSLW = (query: string) => {
//   const eligibleCourt = Array.isArray(query.match(/SGCA|SGHC/))

//   const yearRegex = /^\[(2\d{3})]/
//   const yearMatch = query.match(yearRegex)
//   const eligibleYear = Array.isArray(yearMatch) && Number.parseInt(yearMatch[1]) >= 2000

//   return eligibleCourt && eligibleYear
// }

// const inSGSC = (query: string) => inSLW(query)

const findCase = (query: string): CaseFinderResult[] => {
  return [
    ...findSGCase(query),
    ...findUKCase(query),
    ...findEUCase(query),
    ...findHKCase(query),
  ]
}

const formatAbbrs = (abbrArray) => abbrArray.map(({ abbr, appendum }) => `${abbr
    .split(``)
    .map(letter =>
      /[a-z]/i.test(letter)
        ? letter+`\\.?`
        : letter,
    ).join(``)
    }${appendum ? appendum : ``}`,
  ).join(`|`)

const findSGCase = (query: string): CaseFinderResult[] => {
  const regex = /\[[12]\d{3}]( \d{1,2})? (sgca(\(i\))?|sghc|sgdc|sgmc|slr(\(r\))?) \d{1,4}/gi
  const matches = [...query.matchAll(regex)]
  if (matches.length > 0) {
    return matches.map((match) => ({
      citation: match[0],
      index: match.index,
      jurisdiction: Constants.JURISDICTIONS.SG.id,
    })).map(c => ({ ...c, type: `case` }))
  }
  return []
}

const findUKCase = (query:string): CaseFinderResult[] => {
  const abbrs = formatAbbrs([
    { abbr: `EWCA`, appendum: `( Civ| Crim)?` },
    { abbr: `EWHC`, appendum: `( Patents)?` },
    { abbr: `UKSC` },
    { abbr: `UKPC` },
    { abbr: `UKHL` },
    { abbr: `AC` },
    { abbr: `Ch`, appendum: `( D)?` },
    { abbr: `QB`, appendum: `(D)?` },
    { abbr: `KB` },
    { abbr: `WLR`, appendum: `( ?\\(D\\))?` },
    { abbr: `All ER`, appendum: `( \\(D\\))?` },
    { abbr: `BCLC` },
    { abbr: `BCC` },
    { abbr: `HL Cas` },
    { abbr: `App Cas` },
    { abbr: `Ld Raym` },
    { abbr: `FSR` },
    { abbr: `ECC` },
    { abbr: `ITCLR` },
    { abbr: `RPC` },
    { abbr: `Ex Rep` },
    { abbr: `ER` },
    { abbr: `Cr App R` },
    { abbr: `ALR` },
    { abbr: `FLR` },
    { abbr: `Hare` },
    { abbr: `H & Tw` },
    { abbr: `EMLR` },
    { abbr: `Fam` },
    { abbr: `Macq` },
    { abbr: `TLR` },
    { abbr: `Ves & B` },
    { abbr: `EngR` },
  ])
  // eslint-disable-next-line unicorn/better-regex
  const yearRegex = new RegExp(/((\[|\()[12]\d{3}(-[12]\d{3})?(\]|\)))/)
  const volumeRegex = new RegExp(/( \d{1,2})?/)
  const pageRegex = new RegExp(/\d{1,4}/)
  const regex = new RegExp(`${yearRegex.source}${volumeRegex.source} (${abbrs}) ${pageRegex.source}`, `gi`)

  const matches = [...query.matchAll(regex)]
  if (matches.length > 0) {
    return matches.map((match) => ({
      citation: match[0],
      index: match.index,
      jurisdiction: Constants.JURISDICTIONS.UK.id,
    })).map(c => ({ ...c, type: `case` }))
  }
  return []
}

export const epoRegex = new RegExp(/[GJT][ _]?\d{1,4}\/\d{1,2}/)
export const cjeuRegex = new RegExp(/[CT]-\d{1,3}\/\d{1,2}/)

const findEUCase = (query: string): CaseFinderResult[] => {
  const regex = new RegExp(`(${epoRegex.source})|(${cjeuRegex.source})`, `gi`)
  const cleanedQuery = query.replace(/â€‘/g, `-`)

  const matches = [...cleanedQuery.matchAll(regex)]
  if(matches.length > 0){
    return matches.map((match) => ({
      citation: match[0],
      court: match[1] ? Constants.COURTS.EU_epo.id : Constants.COURTS.EU_cjeu.id,
      index: match.index,
      jurisdiction: Constants.JURISDICTIONS.EU.id,
    })).map(c => ({ ...c, type: `case` }))
  }
  return []
}

const findHKCase = (query: string): CaseFinderResult[] => {
  const yearRegex = new RegExp(/(([([])[12]\d{3}(-[12]\d{3})?(\)]))/)
  const volumeRegex = new RegExp(/( \d{1,2})?/)
  const pageRegex = new RegExp(/\d{1,4}/)
  const abbrs = formatAbbrs([
    { abbr: `HKCA` },
    { abbr: `HKCFA` },
    { abbr: `HKCFI` },
    { abbr: `HKLRD` },
    { abbr: `HKCFAR` },
    { abbr: `HKC` },
    { abbr: `HKDC` },
    { abbr: `HKFC` },
    { abbr: `HKLdT` },
    { abbr: `HKCT` },
    { abbr: `HKCrC` },
    { abbr: `HKOAT` },
    { abbr: `HKLaT` },
    { abbr: `HKMagC` },
  ])
  const regex = new RegExp(`${yearRegex.source}${volumeRegex.source} (${abbrs}) ${pageRegex.source}`, `gi`)

  const matches = [...query.matchAll(regex)]
  if (matches.length > 0) {
    return matches.map((match) => ({
      citation: match[0],
      index: match.index,
      jurisdiction: Constants.JURISDICTIONS.HK.id,
    })).map(c => ({ ...c, type: `case` }))
  }
  return []
}

const CaseFinder = {
  findCase,
  findEUCase,
  findHKCase,
  findSGCase,
  findUKCase,
}

export default CaseFinder