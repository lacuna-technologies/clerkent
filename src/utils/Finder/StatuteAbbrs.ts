import Constants from "../Constants"

const SGAbbrs = [
  {
    abbrs: [`pofma`, `pofma 2019`, `pofma 19`, `fake news law`],
    name: `Protection from Online Falsehoods and Manipulation Act 2019`,
  },
  {
    abbrs: [`cntma`, `c19tma`],
    name: `COVID-19 (Temporary Measures) Act 2020`,
  },
  {
    abbrs: [`ta`, `telecoms act`],
    name: `Telecommunications Act`,
  },
  {
    abbrs: [`bcisopa`, `sop`],
    name: `Building and Construction Industry Security of Payment Act`,
  },
  {
    abbrs: [`bcisopr`, `sop`],
    name: `Building and Construction Industry Security of Payment Regulations`,
  },
  {
    abbrs: [`pc`],
    name: `Penal Code`,
  },
  {
    abbrs: [`cpc`],
    name: `Criminal Procedure Code`,
  },
  {
    abbrs: [`mda`],
    name: `Misuse of Drugs Act`,
  },
  {
    abbrs: [`da`],
    name: `Defamation Act`,
  },
  {
    abbrs: [`clpa`],
    name: `Conveyancing and Law of Property Act`,
  },
  {
    abbrs: [`lpa`],
    name: `Legal Profession Act`,
  },
  {
    abbrs: [`lpcr`, `lp(pc)r`, `lp(pc)r`, `lppcr`],
    name: `Legal Profession (Professional Conduct) Rules`,
  },
].map(abbr => ({ ...abbr, jurisdiction: Constants.JURISDICTIONS.SG.id }))

const UKAbbrs = [
  {
    abbrs: [`lpa`, `lpa 1925`],
    name: `Law of Property Act 1925`,
  },
  {
    abbrs: [`lpmpa`, `lp(mp)a`, `lm(mp)a 1989`],
    name: `Law of Property (Miscellaneous Provisions) Act 1989`,
  },
  {
    abbrs: [`dpa`, `dpa 1988`],
    name: `Data Protection Act 1988`,
  },
  {
    abbrs: [`ea`, `ea 2010`],
    name: `Equality Act 2010`,
  },
  {
    abbrs: [`hra`, `hra 1998`],
    name: `Human Rights Act 1998`,
  },
  {
    abbrs: [`ca`, `ca 2006`],
    name: `Companies Act 2006`,
  },
  {
    abbrs: [`hswa`, `hswa 1974`],
    name: `Health and Safety at Work etc. Act 1974`,
  },
  {
    abbrs: [`sga`, `sga 1979`],
    name: `Sale of Goods Act 1979`,
  },
  {
    abbrs: [`foia`, `foia 2000`],
    name: `Freedom of Information Act 2000`,
  },
  {
    abbrs: [`euwa`, `euwa 2020`, `euwaa`, `euwaa 2020`, `eu(wa)`, `eu(wa) 2020`, `eu(wa)a`, `eu(wa)a 2020`],
    name: `European Union (Withdrawal Agreement) Act 2020`,
  },
  {
    abbrs: [`ipa`, `ipa 2016`],
    name: `Investigatory Powers Act 2016`,
  },
  {
    abbrs: [`eqa`, `eqa 2010`],
    name: `Equality Act 2010`,
  },
  {
    abbrs: [`lra`, `lra 2002`],
    name: `Land Registration Act 2002`,
  },
  {
    abbrs: [`lra`, `lra 1997`],
    name: `Land Registration Act 1997`,
  },
  {
    abbrs: [`pa`, `pa 2004`],
    name: `Patents Act 2004`,
  },
  {
    abbrs: [`cdpa`, `cdpa 1988`],
    name: `Copyright, Designs and Patents Act 1988`,
  },
].map(abbr => ({ ...abbr, jurisdiction: Constants.JURISDICTIONS.UK.id }))

const StatuteAbbrs = [
  ...SGAbbrs,
  ...UKAbbrs,
]

export default StatuteAbbrs