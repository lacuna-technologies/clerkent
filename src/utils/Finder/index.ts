export { default } from './Finder'
export type {
  FinderResult,
  LegislationFinderResult,
  CaseFinderResult,
} from './Finder'
