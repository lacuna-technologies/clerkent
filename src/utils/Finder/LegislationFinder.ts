import Law from "../../types/Law"
import StatuteAbbrs from './StatuteAbbrs'

export interface LegislationFinderResult {
  provisionType: string,
  provisionNumber: string,
  statute: string,
  type: `legislation`,
  jurisdiction: Law.JursidictionCode
}

const unabbreviateStatute = (abbrStatute: string) => {
  const isMatch = StatuteAbbrs.filter(abbr => abbr.abbrs.includes(abbrStatute.trim().toLowerCase()))

  if(isMatch.length === 0){
    return [{ name: abbrStatute }]
  }
  return isMatch.map(({ name, jurisdiction }) => ({ jurisdiction, name }))
}

const provisionAbbreviations = [
  {
    abbrs: [`s`, `sec`, `section`],
    name: `Section`,
  },
  {
    abbrs: [`r`, `rule`],
    name: `Rule`,
  },
  {
    abbrs: [`art`, `article`],
    name: `Article`,
  },
  {
    abbrs: [`o`, `order`],
    name: `Order`,
  },
  {
    abbrs: [`reg`, `regulation`],
    name: `Regulation`,
  },
]
const provisionTypeRegex = provisionAbbreviations.map((a) => a.abbrs.join(`|`)).join(`|`)
const provisionSubsectionRegex = new RegExp(`\\(\\d{1,2}[A-Z]{0,2}\\)`)

const unabbreviateProvision = (provisionType: string) => {
  const isMatch = provisionAbbreviations.filter(({ abbrs }) => abbrs.includes(provisionType.trim().toLowerCase()))
  if(isMatch.length === 1){
    return isMatch[0].name
  }
  return provisionType
}

const findLegislation = (citation: string): LegislationFinderResult[] => {
  // eslint-disable-next-line unicorn/better-regex
  const regex = new RegExp(`((?<provision>(${provisionTypeRegex}) ?\\d{1,4}(${provisionSubsectionRegex.source})*)[ ,]*)?(?<statute>[A-Z]{2,}[ ,A-Z]*( ?[12]\\d{3})?)`, `gi`)
  const cleanedCitation = citation.trim()
  const matches = [...cleanedCitation.matchAll(regex)]
    .map(([_1, _2, provision, _3, _4, statute]) => {
      if(!provision){
        return {
          provisionNumber: false,
          provisionType: false,
          statute,
          type: `legislation`,
        }
      }
      const [_, provisionType, provisionNumber] = provision.match(/([a-z]+) ?(\d+[a-z]*)/i)
      return {
        provisionNumber,
        provisionType: unabbreviateProvision(provisionType),
        statute,
        type: `legislation`,
      }
    })

  return matches.reduce((current, { statute, ...attributes }) => {
    return [
      ...current,
      ...unabbreviateStatute(statute)
        .map(({ name, jurisdiction }) => ({
          ...attributes,
          ...(jurisdiction ? { jurisdiction } : {}),
          statute: name,
        })),
    ]
  }, [])
}

const LegislationFinder = {
  findLegislation,
}

export default LegislationFinder