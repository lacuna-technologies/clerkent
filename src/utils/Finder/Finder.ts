import CaseFinder from './CaseFinder'
import LegislationFinder from './LegislationFinder'

const find = (citation: string) => {
  const legislation = [...LegislationFinder.findLegislation(citation)]
  if(legislation.length > 0){
    return legislation
  }
  return [...CaseFinder.findCase(citation)]
}

const Finder = {
  find,
  findCase: CaseFinder.findCase,
  findLegislation: LegislationFinder.findLegislation,
}

export default Finder