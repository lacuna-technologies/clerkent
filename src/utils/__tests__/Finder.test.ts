import Finder from '../Finder'

describe(`Finder`, () => {
  it(`ðŸ‡¸ðŸ‡¬ citations`, () => {
    const citations = [
      `[2021] SGCA 1`,
      `[2020] SGHC 2`,
    ]

    for(const citation of citations){
      expect(Finder.findSGCase(citation)).toMatchSnapshot()
    }
  })
})