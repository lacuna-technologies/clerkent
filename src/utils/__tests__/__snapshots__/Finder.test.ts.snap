// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Finder ðŸ‡¸ðŸ‡¬ citations 1`] = `
Array [
  Object {
    "citation": "[2021] SGCA 1",
    "index": 0,
    "jurisdiction": "SG",
  },
]
`;

exports[`Finder ðŸ‡¸ðŸ‡¬ citations 2`] = `
Array [
  Object {
    "citation": "[2020] SGHC 2",
    "index": 0,
    "jurisdiction": "SG",
  },
]
`;
